src/main/java/me/karakelley/http/RequestParser.java

```diff
@@ -0,0 +1,62 @@
+package me.karakelley.http;
+
+import me.karakelley.http.exceptions.InvalidRequestException;
+import me.karakelley.http.utility.LineReader;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class RequestParser {
+  private final RequestValidator requestValidator;
+  private Map<String, String> headers = new HashMap<>();
+  private String requestMethod;
+  private String requestPath;
+  private String requestProtocol;
+
+  public RequestParser(RequestValidator requestValidator) {
+    this.requestValidator = requestValidator;
+  }
+
+  public RequestParser parse(LineReader reader) {
+    parseRequestLine(reader);
+    parseHeaders(reader);
+    return this;
+  }
+
+  public Request buildRequest(int port) throws InvalidRequestException {
+    Request request = RequestBuilder.build(requestMethod, requestPath, requestProtocol, headers, port);
+
+    if (requestValidator.isValid(request)) {
+      return request;
+    } else throw new InvalidRequestException("Invalid request");
+  }
+
+  private void parseHeaders(LineReader reader) {
+    try {
+      String lines;
+      while ((lines = reader.readLine()) != null && !lines.isEmpty()) {
+        String[] splitHeaders = lines.split(":", 2);
+        headers.put(splitHeaders[0], splitHeaders[1].trim());
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void parseRequestLine(LineReader reader) {
+    String requestLine;
+    String[] parsedLine;
+    try {
+      requestLine = reader.readLine();
+      parsedLine = requestLine.replaceAll("\\s+", " ").split(" ");
+      if (parsedLine.length == 3) {
```

Comment:
The `requestMethod`, `requestPath` and `requestProtocol` remain null and soon after an `InvalidRequestException` is thrown. It could be thrown here now that I think of it. I will note that most of this parsing functionality has been changed in the `post-request` story but it still might be a good time to throw the exception at this point.

Reply:
I'm willing to leave it alone here and see how you change it in as a result of POST functionality